### ISSUE TITLE
[BUG] Scalars leaking GPU memory

### ISSUE THREAD

**Issue by revans2** [rapidsai/cudf#4396]
**Describe the bug**
When ever I use a scalar some memory leaks because the destructors for scalar are not virtual.

**Steps/Code to reproduce bug**
```
{
  std::unique_ptr<cudf::scalar> s = cudf::make_numeric_scalar(cudf::data_type(cudf::BOOL8));
}
```

512 bytes of GPU memory were just leaked, which in real world situations causes a lot of issues with fragmentation.

**Expected behavior**
No GPU memory is leaked.

**Comment by jrhemstad** [comment]
Yep, this was a major oversight. Should be a trivial fix though.

**Comment by kkraus14** [comment]
Fixed by #4399
